### PR TITLE
Upload artifacts showing diff images on test failure

### PR DESCRIPTION
**Pull request by weiji14**
**Description of proposed changes**

When our unit tests fail, it is useful sometimes to see what went wrong. Specifically, tests with `pytest-mpl.mpl_image_compare` and `check_figures_equal` (#555) will return an image diff on test failures in the 'tmp-test-dir-with-unique-name' directory, and we can upload those files as a [Github artifact](https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#artifacts) for inspection.

Uses the [Upload Artifact](https://github.com/actions/upload-artifact) Github Action.

Originally suggested at https://github.com/GenericMappingTools/pygmt/pull/471#discussion_r512388019

Fixes #

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Notes**

- You can write `/format` in the first line of a comment to lint the code automatically

### CODE REVIEW

File: pygmt/tests/test_grdimage.py [hunk 1]
```diff
@@ -154,7 +154,7 @@ def test_grdimage_central_meridians(grid, proj_type, lon0):
 # are still slightly different with an RMS error of 25, see issue at
 # https://github.com/GenericMappingTools/pygmt/issues/390
 # TO-DO remove tol=1.5 and pytest.mark.xfail once bug is solved in upstream GMT
-@check_figures_equal(tol=1.5)
+# @check_figures_equal(tol=1.5)
 @pytest.mark.parametrize("lat0", [0, 30])
 @pytest.mark.parametrize("lon0", [0, 123, 180])
 @pytest.mark.parametrize("proj_type", [pytest.param("Q", marks=pytest.mark.xfail), "S"])
```
Comment (weiji14) [inline pygmt/tests/test_grdimage.py:160]: Why are the `test_grdimage_central_meridians_and_standard_parallels` all XPASS now?!! Arent we still using GMT 6.1.1? Does this mean #390 is magically resolved somehow?

```
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-0-0] XPASS [ 53%]
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-0-30] XPASS [ 53%]
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-123-0] XPASS [ 54%]
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-123-30] XPASS [ 54%]
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-180-0] XPASS [ 54%]
..\tests\test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-180-30] XPASS [ 55%]
```

Comment (seisman) [reply pygmt/tests/test_grdimage.py:160]: That's weird. The master branch CI says these tests are xfail:
```
../tests/test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-123-0-png] XFAIL [ 52%]
../tests/test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-123-30-png] XFAIL [ 53%]
../tests/test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-180-0-png] XFAIL [ 53%]
../tests/test_grdimage.py::test_grdimage_central_meridians_and_standard_parallels[Q-180-30-png] XFAIL [ 53%]
```

File: .github/workflows/ci_tests.yaml [hunk 2]
```diff
@@ -117,6 +117,14 @@ jobs:
         shell: bash -l {0}
         run: make test PYTEST_EXTRA="-r P"
 
+      # Upload diff images on test failure
+      - name: Upload diff images if any test fails
+        uses: actions/upload-artifact@v2
+        if: ${{ failure() }}
+        with:
+          name: artifact-${{ matrix.os }}-${{ matrix.python-version }}
```
Comment (seisman) [inline .github/workflows/ci_tests.yaml:125]: It works as expected now. 

image

What about changing the names to:
```suggestion
          name: artifact-${{ runner.os }}-Python${{ matrix.python-version }}
```

[`runner.os`](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#runner-context)'s possible values are Linux, Windows, or macOS. 

Then the artifacts will be `artifact-Windows-Python3.8`.

Comment (weiji14) [reply .github/workflows/ci_tests.yaml:125]: > [`runner.os`](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#runner-context)'s possible values are Linux, Windows, or macOS.
> 
> Then the artifacts will be `artifact-Windows-Python3.8`.

Oh yes, shorter name is good! I'll also add this to the GMT Latest tests yaml file later.

File: .github/workflows/ci_tests_dev.yaml [hunk 3]
```diff
@@ -86,3 +86,11 @@ jobs:
       # Run the tests
       - name: Test with pytest
         run: make test PYTEST_EXTRA="-r P"
+
+      # Upload diff images on test failure
+      - name: Upload diff images if any test fails
+        uses: actions/upload-artifact@v2
+        if: ${{ failure() }}
+        with:
+          name: artifact-GMT ${{ matrix.gmt_git_ref }}-${{ runner.os }}
```
Comment (seisman) [inline .github/workflows/ci_tests_dev.yaml:95]:
```suggestion
          name: artifact-GMT${{ matrix.gmt_git_ref }}-${{ runner.os }}
```
or
```suggestion
          name: artifact-GMT-${{ matrix.gmt_git_ref }}-${{ runner.os }}
```

File: pygmt/tests/test_grdimage.py [hunk 4]
```diff
@@ -154,7 +154,7 @@ def test_grdimage_central_meridians(grid, proj_type, lon0):
 # are still slightly different with an RMS error of 25, see issue at
 # https://github.com/GenericMappingTools/pygmt/issues/390
 # TO-DO remove tol=1.5 and pytest.mark.xfail once bug is solved in upstream GMT
-@check_figures_equal(tol=1.5)
+# @check_figures_equal(tol=1.5)
```
Comment (seisman) [inline pygmt/tests/test_grdimage.py:157]: Remember to remove the changes before merging.

**Review by seisman — APPROVED** [review]
Looks good, just a few more comments.

File: pygmt/tests/test_grdimage.py (hunk 1 above)
Comment (weiji14) [reply pygmt/tests/test_grdimage.py:160]: GMT Latest/master CI tests are all XPASS for those on Linux and macOS (see e.g. https://github.com/GenericMappingTools/pygmt/runs/1313952289#step:10:222) :confused:

File: .github/workflows/ci_tests.yaml (hunk 2 above)
Comment (weiji14) [reply .github/workflows/ci_tests.yaml]: Alright, works on GMT Latest/master tests as well:

image

**Review by seisman — APPROVED** [review]

**Pull request merged**